source/opt/eliminate_dead_input_components_pass.cpp

```diff
@@ -80,10 +86,20 @@ Pass::Status EliminateDeadInputComponentsPass::Process() {
     unsigned max_idx = FindMaxIndex(var, original_max);
     if (max_idx != original_max) {
       ChangeStructLength(var, max_idx + 1);
+      vars_to_move.push_back(&var);
       modified = true;
     }
   }
 
+  // Move changed vars after their new type instruction to preserve backward
+  // referencing
+  for (auto var : vars_to_move) {
+    auto type_id = var->type_id();
+    auto type_inst = def_use_mgr->GetDef(type_id);
+    var->RemoveFromList();
+    var->InsertAfter(type_inst);
```

Comment:
If I follow this:
 - we have a type instruction somewhere in the code, and somewhere down the line, the variable instruction.
 - we create a new type (resized array or resized struct), which is inserted somewhere, without constraint on location.
 - we move the variable declaration from it's original stop to after this new type.
 
 Since we are only dealing with input/output variables, I guess they are on the global scope, so their position shouldn't change much, but not 100% sure. Could their position change cause something?
Also, why are we changing the variable declaration position, but not the type instruction to be just below the old type? (Since it's kept until the next DCE pass no?)

Reply:
I created the new type through the type manager which will make sure that all the type instructions are sequenced correctly. That is the beauty of a type manager. But I still have to make sure the variable instruction is sequenced correctly relative to its type.

The type is the only think we need to worry about here. The only other field in OpVariable with an id is the optional initializer, but input and output don't take initializers.